### PR TITLE
Feature/entity management alert

### PR DESCRIPTION
**Pull request by Ayoush**
`Problem Statement` :- To provide support of alert in entity management system.

`In depth explanation` :- Provided and explained more clearly in this [doc](https://datakrew.atlassian.net/wiki/spaces/AN/pages/edit-v2/110985223?draftShareId=c6415362-193c-400e-9712-75777595b49c)

`Changes` :-
- [x] Alert Rule Migration  
- [x] Alert Rule Schema
- [x] Alert Rule Controller
- [x] Partials Schema
- [x] Alert Creation 
- [x] Alert Generation
- [x] Taking sensors data from queue
- [x] Pushing generated alert data into alert queue
- [x] Alert generation
- [x] Testcases of Alert Rule Schema
- [x] Testcases of Alert Rule endpoints
- [x] Testcases of Alert Creation

**NOTE**
**Needs to set ENV variables for the following keys as config to BROADWAY:**
- **BROADWAY_CONN_USERNAME**
- **BROADWAY_CONN_PASSWORD**
- **BROADWAY_CONN_HOST**

### CODE REVIEW

File: apps/acqdat_core/lib/acqdat_core/entity-management/model/grouping.ex [hunk 1]
```diff
@@ -0,0 +1,201 @@
+defmodule AcqdatCore.EntityManagement.Model.Grouping do
+  @moduledoc """
+  A module which takes care of the grouping logic for the alerts.
+
+  Grouping is an important feature for alerts which avoids too many notifications
+  being sent to a user for any event.
+
+  All the alerts which are generated from the same entity(sensor, gateway, asset)
```
Comment (arjun289) [inline apps/acqdat_core/lib/acqdat_core/entity-management/model/grouping.ex:8]: Why have you commented out all the code in this module.

**Review by arjun289 — CHANGES_REQUESTED** [review]

File: apps/acqdat_api/lib/acqdat_api_web/router.ex [hunk 2]
```diff
@@ -267,6 +267,18 @@ defmodule AcqdatApiWeb.Router do
...
+      scope "/" do
+        get "/policies", AlertRulesController, :policies
+        get "/grouping_rules", AlertRulesController, :grouping_rules
+        resources "/alert-rules", AlertRulesController,
+          except: [:new, :edit],
+          as: :entity_alert_rule
+
+        # get "/alert_rule_listing", AlertFilterListingController, :alert_rule_listing
+        # get "/alert_apps", AlertFilterListingController, :alert_app_listing
+        # get "/alert_status", AlertFilterListingController, :alert_status_listing
```
[the remark this answers is not in the document]
Comment (BandanaPandey) [reply apps/acqdat_api/lib/acqdat_api_web/router.ex]: done

File: apps/acqdat_core/lib/acqdat_core/entity-management/alert_creations/alert_creation.ex [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+    Broadway.start_link(__MODULE__,
+      name: __MODULE__,
+      producer: [
+        module:
+          {BroadwayRabbitMQ.Producer,
+           queue: @entity_queue,
+           declare: [durable: true],
+           connection: [
+             username: "guest",
+             password: "guest"
```
[the remark this answers is not in the document]
Comment (BandanaPandey) [reply apps/acqdat_core/lib/acqdat_core/entity-management/alert_creations/alert_creation.ex]: done

File: apps/acqdat_api/lib/acqdat_api/entity-management/alert_rules.ex [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  def create(params) do
+    params = params_extraction(params)
+    verify_alert_rules(AlertRules.create(params))
+  end
+
+  def verify_alert_rules({:ok, alert_rules}) do
+    {:ok, alert_rules}
+  end
+
+  def verify_alert_rules({:error, message}) do
```
[the remark this answers is not in the document]
Comment (BandanaPandey) [reply apps/acqdat_api/lib/acqdat_api/entity-management/alert_rules.ex]: done

File: apps/acqdat_core/lib/acqdat_core/iot-manager/data_handler/data_parser.ex [hunk 5]
```diff
@@ -74,11 +88,27 @@ defmodule AcqdatCore.IotManager.DataParser do
...
+
+    with :ok <- AMQP.Exchange.declare(channel, @exchange, :fanout, durable: true),
+         {:ok, _} <- AMQP.Queue.declare(channel, @entity_queue, durable: true),
+         :ok <- AMQP.Queue.bind(channel, @entity_queue, @exchange),
+         {:ok, _} <- AMQP.Queue.declare(channel, @alert_queue, durable: true),
+         :ok <- AMQP.Queue.bind(channel, @alert_queue, @exchange) do
+      AMQP.Basic.publish(channel, @exchange, "", Jason.encode!(data))
+    end
+
+    AMQP.Connection.close(connection)
```
Comment (Ayoush) [inline apps/acqdat_core/lib/acqdat_core/iot-manager/data_handler/data_parser.ex:109]: Connection should not be closed

**Pull request merged**